Dockerfile

```diff
@@ -1,22 +1,45 @@
-FROM python:3.7-slim-stretch
+FROM node:current-alpine AS NODEBUILD
+# copy directory to /opt and work inside it
+WORKDIR /opt
+COPY . /opt
+# run the npm portion of build-static script
+RUN rm -rf static && mkdir -p static/
+RUN npm install -g npm && \
+    npm install && \
+    npm run build
 
-# Install packages required for uswgi
-RUN apt-get update
-RUN apt-get -y install build-essential
-RUN apt-get -y install mime-support
 
-# Install requirements and uwsgi server for running python web apps
+FROM python:3.9-slim AS PYTHONBUILD
+# install missing system deps
+RUN apt-get update && apt-get -y install build-essential mime-support libpcre3-dev libsass-dev mime-support
+# set working dir
 WORKDIR /etc/linkding
 COPY requirements.prod.txt ./requirements.txt
-RUN pip install -U pip
-RUN pip install -Ur requirements.txt
+# make a virtualenv, upgrade pip & wheel, install prod requirements
+RUN mkdir /venv && \
+    python -m venv --upgrade-deps --copies /venv && \
+    /venv/bin/pip install --upgrade pip wheel && \
+    /venv/bin/pip install -Ur requirements.txt
+# run the python half of build-static script
+COPY --from=NODEBUILD /opt/ .
+RUN /venv/bin/python manage.py compilescss && \
+    /venv/bin/python manage.py collectstatic --ignore=*.scss && \
+    /venv/bin/python manage.py compilescss --delete-files
 
-# Copy application
-COPY . .
 
-# Expose uwsgi server at port 9090
+FROM python:3.9-slim as FINAL
+WORKDIR /etc/linkding
+# copy application code from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /etc/linkding  /etc/linkding
+# copy virtualenv from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /venv /venv
+# copy mime.types from PYTHONBUILD stage
+COPY --from=PYTHONBUILD /etc/mime.types /etc/mime.types
+# run bootstrap.sh
+RUN mkdir -p data && \
+    /venv/bin/python manage.py migrate && \
+    /venv/bin/python manage.py generate_secret_key && \
+    chown -R www-data: /etc/linkding/data
```

Comment:
All of these need to run when the container is created / after the data folder is mounted into the container:
- `/venv/bin/python manage.py migrate` the purpose is to upgrade the database of the installation if the schema changes, which means it needs access to the respective database
- `/venv/bin/python manage.py generate_secret_key` the secret key must be unique per linkding installation, with this change every install would use the same secret
- same for chown I think - this was to guard against permission issues when mounting a data folder into the container

So all of these would need to run as part of CMD. I guess the idea was to unify the Docker setup a bit, but with the restrictions above I think keeping the separate (and well documented) bootstrap.sh is the better trade-off.

Reply:
Ah, I totally forgot about the django migrations and django-generate-secret-key were idempotent, so you could just replace this with the `bootstrap.sh` script. 
As for the migrations, `/venv/bin/python manage.py migrate` should create the database if the sqlite3 file is missing. But that does require the user deploying the container to run the container mounting the `/etc/linkding/data` directory.